calyx/src/passes/top_down_static_timing.rs

```diff
@@ -134,16 +326,85 @@ fn enable_calculate_states(
     let mut assigns = build_assignments!(builder;
         group["go"] = pre_guard ? signal_on["out"];
     );
+
+    // Enable when in range of group's latency.
     schedule
         .enables
         .entry(range)
         .or_default()
         .append(&mut assigns);
-    cur_state + time
+
+    // Add any necessary internal transitions. In the case time is 1 and there
+    // is a single transition, it is taken care of by the parent.
+    let starts = cur_state..cur_state + time - 1;
+    let ends = cur_state + 1..cur_state + time;
+    schedule
+        .transitions
+        .extend(starts.zip(ends).map(|(s, e)| (s, e, pre_guard.clone())));
+
+    Ok(vec![(cur_state + time, pre_guard.clone())])
 }
 
-#[derive(Default)]
-pub struct TopDownStaticTiming;
+/// Helper to assert attributes have a static annotation, and return its latency.
+fn get_static_time(attributes: &Attributes) -> u64 {
```

Comment:
I wasn't sure, do we want to assert all nodes have a static attribute, or does it only matter for enables? I haven't looked into how that attribute is inferred, but from what I've tested, if all the enables are static, the whole control program becomes static.

Reply:
That makes sense, and I was planning to assert this everywhere to expect `infer-static-timing` had been run. I'm curious if that can be relaxed though... unless I'm missing something, the current implementation only really needs to have the static attribute on enables. Less requirements seems like a good thing if that is sufficient.

Reply:
Now that I'm starting to think about how this slots in for tests, I'm curious what the best approach is. Should this pass just ignore IR it can't handle, and leave it to tdcc to deal with later? Or should we defensively error out and expect this pass to not be enabled in situations it can't work?

Reply:
A couple more points here: apparently while constructs don't receive a @static annotation even when all children do. In turn, this causes a seq generated by remove-comb-groups to not receive a @static annotation. So I think it makes sense to only enforce this on enables. The question remains, what if all enables are not @static, should it error out or do nothing?

Reply:
While nodes receive a static annotation when they have a @bound annotation on them and all their children have @static annotation too. Without bound we don’t know how many times the loop executes.

Reply:
This pass is considered an optimization and should in general bail out when it can’t do anything. However, for tests, it makes sense to provide a flag that throws an error instead of silently passing.

Reply:
Also yes you are correct about only enables needing static annotations. Infer static timing propagates them to parents do that you don’t have to traverse the tree to decide if you can compile a control sub program. Either option is okay with me so up to you.

Reply:
If this is just an optimization then I will make it do nothing if the whole progam isn't @static. I'm not sure yet where to get the @bound attribute. Is there anything to infer that, or should it be supplied from the frontend?

Reply:
Generally frontends are expected to provide it. We don’t have a pass to infer it but could be cool…

Reply:
Ok, I will make sure we stamp it out from CIRCT; it's known statically in every program we support pipelining.